esp32-firmware/src/main.cpp

```
@@ -314,20 +459,21 @@ void processingTask(void * pvParameters) {
+        // Add raw waveform data at the full sampling rate. Export only two periods
+        // to keep MQTT payload bounded while improving chart detail.
         JsonArray waveV_arr = doc["waveform_v"].to<JsonArray>();
         JsonArray waveI_arr = doc["waveform_i"].to<JsonArray>();
-        int samplesPerCycle = round(SAMPLING_FREQ / freq);  // Use detected frequency, not hardcoded 50Hz
-        // Send 1 full cycle + 1 extra sample to complete the period visually
-        int samplesToSend = 3 * samplesPerCycle + 1;
-        if (samplesToSend > SAMPLES) samplesToSend = SAMPLES;
+        int samplesPerCycle = round(SAMPLING_FREQ / freq);  // Use detected frequency, not hardcoded 50 Hz
+        int samplesToSend = WAVEFORM_EXPORT_PERIODS * samplesPerCycle + 1;
+        int maxSamplesToSend = SAMPLES;
+        if (samplesToSend > maxSamplesToSend) samplesToSend = maxSamplesToSend;
```

Clang (14.0.6)

[note] +4, including nesting penalty of 3, nesting level increased to 4

(clang)